### ISSUE TITLE
nvme_pcicfg_map_bar fails on ARM64 with vfio-pci and IOMMU enabled

### ISSUE THREAD

**Issue by ksingh2017** [spdk/spdk#226]
This issue looks somewhat similar to the one below but I'm not sure whether they're related.  After every soft reboot I do see /dev/nvme0n1.  
(Probe fails and device does not show up anymore, even after reboot, after resetting to nvme.  #105).

SPDK is unable to locate Nvme0n1.  After some debug I noticed that lsblk doesn't show /dev/nvme0n1 after spdk/scripts/setup.sh was run.  

Used SPDK commit c3b313811761956dde1d5a2df50b35eec178955d
Author: Xiaodong Liu <xiaodong.liu@intel.com>
Date:   Tue Dec 12 04:53:18 2017 -0500

Compile SPDK on ARM with CONFIG_RDMA=y CONFIG_DEBUG=y

# ./setup.sh 
0006:58:00.0 (8086 0953): nvme -> vfio-pci

Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid13531 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
invalid usermode virtual address 0xffff40000000
invalid usermode virtual address 0xfffee0000000
app.c: 369:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 668:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
EAL:   using IOMMU type 1 (Type 1)
EAL: Releasing pci mapped resource for 0006:58:00.0
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: *NOTICE*: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: *NOTICE*: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 400:spdk_nvmf_construct_subsystem: *ERROR*: Could not find namespace bdev 'Nvme0n1'
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: *NOTICE*: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: *NOTICE*: Acceptor running

**Comment by benlwalker** [comment]
When you run setup.sh it unloads the kernel driver so that our user space driver can be loaded instead. That means it is going to disappear from the /dev directory. That's expected behavior. 

Your issue is probably a missing or incorrect [Nvme] section in your config file.

**Comment by ksingh2017** [comment]
Thanks Ben!  You were partially right.  In trying to debug this issue, I had copied the nvmf_tgt script to another system.  I've updated the [Nvme] section to point to the correct disk but am still seeing the issue.  dmesg shows that vfio failed to setup IOMMU DMA ops.  I've provided some debug info below.  Any suggestions?  

BTW, DPDK compile was failing for missing numa.h and numaif.h so I had to disable these config settings (CONFIG_RTE_EAL_NUMA_AWARE_HUGEPAGES=n and CONFIG_RTE_LIBRTE_VHOST_NUMA=n).

==================================================
[Nvme]
  TransportId "trtype:PCIe traddr:0006:58:00.0" Nvme0
  Timeout 0
  ActionOnTimeout None
  AdminPollRate 100000
  HotplugEnable No
==================================================
Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid3585 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
invalid usermode virtual address 0xffff20000000
app.c: 369:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 668:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
EAL:   using IOMMU type 1 (Type 1)
nvme_pcie.c: 548:nvme_pcie_ctrlr_allocate_bars: *ERROR*: nvme_pcicfg_map_bar failed with rc 0 or bar (nil)
nvme.c: 352:nvme_ctrlr_probe: *ERROR*: Failed to construct NVMe controller
EAL: Releasing pci mapped resource for 0006:58:00.0
EAL: Requested device 0006:58:00.0 cannot be used
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: *NOTICE*: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: *NOTICE*: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 400:spdk_nvmf_construct_subsystem: *ERROR*: Could not find namespace bdev 'Nvme0n1'
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: *NOTICE*: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: *NOTICE*: Acceptor running
==================================================
# dmesg
[  277.786845] VFIO - User Level meta-driver version: 0.3
[  280.143637] Failed to set up IOMMU for device 0006:58:00.0; retaining platform DMA ops
[  281.769424] pci_bus 0000:01: busn_res: [bus 01] end is updated to 01
[  281.769436] pci_bus 0000:02: busn_res: [bus 02] end is updated to 02
[  281.769445] pci_bus 0000:03: busn_res: [bus 03] end is updated to 03
[  281.769454] pci_bus 0000:04: busn_res: [bus 04] end is updated to 04
[  281.769503] pci 0000:00:01.0: PCI bridge to [bus 01]
[  281.774493] pci 0000:00:14.0: PCI bridge to [bus 02]
[  281.779455] pci 0000:00:15.0: PCI bridge to [bus 03]
[  281.784434] pci 0000:00:16.0: PCI bridge to [bus 04]
[  281.789445] pci_bus 0002:01: busn_res: [bus 01] end is updated to 01
[  281.789454] pci 0002:00:02.0: PCI bridge to [bus 01]
[  281.794544] pci_bus 0004:21: busn_res: [bus 21] end is updated to 21
[  281.794549] pci_bus 0004:20: busn_res: [bus 20-21] end is updated to 21
[  281.794612] pci_bus 0005:90: busn_res: [bus 90] end is updated to 90
[  281.794735] pci_bus 0006:58: busn_res: [bus 58] end is updated to 58
[  284.802751] vfio_ecap_init: 0006:58:00.0 hiding ecap 0x19@0x2a0
==================================================
# lspci -s 58:0.0 -vv
0006:58:00.0 Non-Volatile memory controller: Intel Corporation PCIe Data Center SSD (rev 01) (prog-if 02 [NVM Express])
        Subsystem: Intel Corporation SSD 750 Series [Add-in Card]
        Control: I/O- Mem- BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Interrupt: pin A routed to IRQ 25
        NUMA node: 0
        Region 0: Memory at 895010010000 (64-bit, non-prefetchable) [disabled] [size=16K]
        Expansion ROM at 895010000000 [disabled] [size=64K]
        Capabilities: [40] Power Management version 3
                Flags: PMEClk- DSI- D1- D2- AuxCurrent=0mA PME(D0-,D1-,D2-,D3hot-,D3cold-)
                Status: D3 NoSoftRst+ PME-Enable- DSel=0 DScale=0 PME-
        Capabilities: [50] MSI-X: Enable- Count=32 Masked-
                Vector table: BAR=0 offset=00002000
                PBA: BAR=0 offset=00003000
        Capabilities: [60] Express (v2) Endpoint, MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0, Latency L0s <4us, L1 <4us
                        ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset+ SlotPowerLimit 0.000W
                DevCtl: Report errors: Correctable- Non-Fatal- Fatal- Unsupported-
                        RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+ FLReset-
                        MaxPayload 256 bytes, MaxReadReq 512 bytes
                DevSta: CorrErr- UncorrErr- FatalErr- UnsuppReq- AuxPwr- TransPend-
                LnkCap: Port #0, Speed 8GT/s, Width x4, ASPM L0s L1, Exit Latency L0s <4us, L1 <4us
                        ClockPM- Surprise- LLActRep- BwNot- ASPMOptComp+
                LnkCtl: ASPM Disabled; RCB 64 bytes Disabled- CommClk-
                        ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
                LnkSta: Speed 8GT/s, Width x4, TrErr- Train- SlotClk- DLActive- BWMgmt- ABWMgmt-
                DevCap2: Completion Timeout: Range ABCD, TimeoutDis+, LTR-, OBFF Not Supported
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis-, LTR-, OBFF Disabled
                LnkCtl2: Target Link Speed: 8GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance De-emphasis: -6dB
                LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete+, EqualizationPhase1+
                         EqualizationPhase2+, EqualizationPhase3+, LinkEqualizationRequest-
        Capabilities: [100 v1] Advanced Error Reporting
                UESta:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UEMsk:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UESvrt: DLP+ SDES+ TLP- FCP+ CmpltTO- CmpltAbrt- UnxCmplt- RxOF+ MalfTLP+ ECRC- UnsupReq- ACSViol-
                CESta:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- NonFatalErr-
                CEMsk:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- NonFatalErr+
                AERCap: First Error Pointer: 00, GenCap+ CGenEn- ChkCap+ ChkEn-
        Capabilities: [150 v1] Virtual Channel
                Caps:   LPEVC=0 RefClk=100ns PATEntryBits=1
                Arb:    Fixed- WRR32- WRR64- WRR128-
                Ctrl:   ArbSelect=Fixed
                Status: InProgress-
                VC0:    Caps:   PATOffset=00 MaxTimeSlots=1 RejSnoopTrans-
                        Arb:    Fixed- WRR32- WRR64- WRR128- TWRR128- WRR256-
                        Ctrl:   Enable+ ID=0 ArbSelect=Fixed TC/VC=ff
                        Status: NegoPending- InProgress-
        Capabilities: [180 v1] Power Budgeting <?>
        Capabilities: [190 v1] Alternative Routing-ID Interpretation (ARI)
                ARICap: MFVC- ACS-, Next Function: 0
                ARICtl: MFVC- ACS-, Function Group: 0
        Capabilities: [270 v1] Device Serial Number 55-cd-2e-40-4b-fa-19-a6
        Capabilities: [2a0 v1] #19
        Kernel driver in use: vfio-pci
        Kernel modules: nvme
==================================================

**Comment by benlwalker** [comment]
The kernel seems to be unable to utilize the IOMMU on your platform. That's a problem outside of SPDK and I don't have any ideas on that one. 

You could try disabling the IOMMU as a workaround. Then SPDK will use uio instead of vfio in that case and it may work.

**Comment by jimharris** [comment]
I think this is a difference in virtual address space allocated between x86 and your ARM platform.

Could you try changing SHIFT_128TB from 47 to 48 in lib/env_dpdk/env_internal.h?  The 0xFFFF200000000000 address shown has bit 47 cleared, which would trigger the error message.

Just changing that value should be enough.  If that works, we can bump the limit from 128TB to 256TB in upstream, and change the #define names to match.

Thanks,

-Jim

**Comment by ksingh2017** [comment]
Jim,

Thanks for that suggestion. I still see the same issue with that change. However, I’d like to redo after a clean build to be sure that the change took effect.

When I build spdk with DPDK_DIR= pointing to dpdk-17.11 (from tarball) it seems like the makefiles under github/spdk/dpdk/ are being invoked in addition to those under github/spdk/dpdk-17.11/. I had to change the .config under both spdk/dpdk and spdk/dpdk-17.11 to avoid numa.h inclusion compilation errors. What are the build/rebuild requirements for spdk and dpdk?

Thanks,
Kamaljit

**Comment by jimharris** [comment]
Hi Kamaljit,

A different DPDK config file is really needed for ARM which would solve this issue.  Currently the same DPDK config files are used for SPDK on any platform.

Please advise once you run the data again after rebuilding.  If it still fails, please forward the debug info to see if it’s now a different error.

Thanks,

-Jim

**Comment by ksingh2017** [comment]
Jim,

After clean build I still see the same behavior, however, I don't see the "invalid usermode virtual address 0xffff20000000" message now.  In your reply you mentioned ffff2000_0000_0000, where did you see that number?  Its not the same as error message I just mentioned.

# SPDK Console Output:
Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid25164 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
app.c: 369:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 668:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
EAL:   using IOMMU type 1 (Type 1)
nvme_pcie.c: 548:nvme_pcie_ctrlr_allocate_bars: ***ERROR***: nvme_pcicfg_map_bar failed with rc 0 or bar (nil)
nvme.c: 352:nvme_ctrlr_probe: *ERROR*: Failed to construct NVMe controller
EAL: Releasing pci mapped resource for 0006:58:00.0
EAL: Requested device 0006:58:00.0 cannot be used
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: *NOTICE*: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: *NOTICE*: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 400:spdk_nvmf_construct_subsystem: *ERROR*: Could not find namespace bdev 'Nvme0n1'
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: *NOTICE*: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: *NOTICE*: Acceptor running

Thanks,
Kamaljit

**Comment by jimharris** [comment]
Hi,

Traveling at the moment so can’t see kernel source - but is this the st
[ 277.786845] VFIO - User Level meta-driver version: 0.3
[ 280.143637] Failed to set up IOMMU for device 0006:58:00.0; retaining platform DMA ops

On Jan 18, 2018, at 4:32 PM, ksingh2017 <notifications@github.com<mailto:notifications@github.com>> wrote:


Jim,

After clean build I still see the same behavior, however, I don't see the "invalid usermode virtual address 0xffff20000000" message now. In your reply you mentioned ffff2000_0000_0000, where did you see that number? Its not the same as error message I just mentioned.

SPDK Console Output:

Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid25164 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
app.c: 369:spdk_app_start: NOTICE: Total cores available: 1
reactor.c: 668:spdk_reactors_init: NOTICE: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: NOTICE: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: NOTICE: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL: probe driver: 8086:953 spdk_nvme
EAL: using IOMMU type 1 (Type 1)
nvme_pcie.c: 548:nvme_pcie_ctrlr_allocate_bars: ERROR: nvme_pcicfg_map_bar failed with rc 0 or bar (nil)
nvme.c: 352:nvme_ctrlr_probe: ERROR: Failed to construct NVMe controller
EAL: Releasing pci mapped resource for 0006:58:00.0
EAL: Requested device 0006:58:00.0 cannot be used
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: NOTICE: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: NOTICE: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: NOTICE: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: NOTICE: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 400:spdk_nvmf_construct_subsystem: ERROR: Could not find namespace bdev 'Nvme0n1'
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: NOTICE: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: NOTICE: Acceptor running

Thanks,
Kamaljit

**Comment by jimharris** [comment]
Argh.  Accidentally hit send.

Is this the standard IOMMU driver or a custom one for this platform?

Also can you describe the PCI topology on this system?

On Jan 18, 2018, at 4:59 PM, Harris, James R <james.r.harris@intel.com<mailto:james.r.harris@intel.com>> wrote:

Hi,

Traveling at the moment so can’t see kernel source - but is this the st
[ 277.786845] VFIO - User Level meta-driver version: 0.3
[ 280.143637] Failed to set up IOMMU for device 0006:58:00.0; retaining platform DMA ops

On Jan 18, 2018, at 4:32 PM, ksingh2017 <notifications@github.com<mailto:notifications@github.com>> wrote:


Jim,

After clean build I still see the same behavior, however, I don't see the "invalid usermode virtual address 0xffff20000000" message now. In your reply you mentioned ffff2000_0000_0000, where did you see that number? Its not the same as error message I just mentioned.

SPDK Console Output:

Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid25164 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
app.c: 369:spdk_app_start: NOTICE: Total cores available: 1
reactor.c: 668:spdk_reactors_init: NOTICE: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: NOTICE: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: NOTICE: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL: probe driver: 8086:953 spdk_nvme
EAL: using IOMMU type 1 (Type 1)
nvme_pcie.c: 548:nvme_pcie_ctrlr_allocate_bars: ERROR: nvme_pcicfg_map_bar failed with rc 0 or bar (nil)
nvme.c: 352:nvme_ctrlr_probe: ERROR: Failed to construct NVMe controller
EAL: Releasing pci mapped resource for 0006:58:00.0
EAL: Requested device 0006:58:00.0 cannot be used
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: NOTICE: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: NOTICE: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: NOTICE: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: NOTICE: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 400:spdk_nvmf_construct_subsystem: ERROR: Could not find namespace bdev 'Nvme0n1'
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: NOTICE: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: NOTICE: Acceptor running

Thanks,
Kamaljit

**Comment by ksingh2017** [comment]
Its the standard IOMMU driver. However, kernel 4.14.8 was installed by another team member and I see these in the .config.  Shouldn't CONFIG_VFIO_NOIOMMU be turned off?

## grep VFIO ./linux-4.14.8/.config
CONFIG_VFIO_IOMMU_TYPE1=m
CONFIG_VFIO_VIRQFD=m
CONFIG_VFIO=m
**CONFIG_VFIO_NOIOMMU=y**
CONFIG_VFIO_PCI=m
CONFIG_VFIO_PCI_MMAP=y
CONFIG_VFIO_PCI_INTX=y
CONFIG_VFIO_PLATFORM=m
CONFIG_VFIO_AMBA=m
# CONFIG_VFIO_PLATFORM_CALXEDAXGMAC_RESET is not set
# CONFIG_VFIO_PLATFORM_AMDXGBE_RESET is not set
# CONFIG_VFIO_MDEV is not set
CONFIG_KVM_VFIO=y

# lspci -t
-+-[0006:57]---00.0-[58]----00.0
 +-[0005:8f]---00.0-[90]--+-00.0
 |                        +-00.1
 |                        +-00.2
 |                        +-00.3
 |                        +-00.4
 |                        \-00.5
 +-[0004:1f]---00.0-[20-21]----00.0-[21]----00.0
 +-[0002:00]-+-02.0-[01]----00.0
 |           \-03.0
 +-[0001:00]-+-08.0
 |           +-09.0
 |           +-0a.0
 |           \-0b.0
 \-[0000:00]-+-01.0-[01]--+-00.0
             |            +-00.1
             |            +-01.3
             |            +-01.4
             |            +-06.0
             |            +-06.1
             |            +-06.2
             |            +-06.3
             |            +-06.4
             |            +-06.5
             |            +-06.6
             |            +-06.7
             |            +-07.0
             |            +-07.1
             |            +-07.2
             |            +-07.3
             |            +-07.4
             |            +-07.5
             |            +-07.6
             |            +-07.7
             |            +-09.2
             |            +-09.4
             |            +-0a.0
             |            +-0a.1
             |            +-0a.2
             |            +-0a.3
             |            +-10.0
             |            \-10.1
             +-09.0
             +-10.0
             +-11.0
             +-14.0-[02]----00.0
             +-15.0-[03]----00.0
             \-16.0-[04]----00.0

**Comment by optimistyzy** [comment]
Will it work with UIO? You may try to bind the device into UIO to see
whether this issue still happens.

2018-01-19 9:28 GMT+08:00 ksingh2017 <notifications@github.com>:

**Comment by jimharris** [comment]
Hi,

Saw your last note on NOIOMMU.

In your debug log below it says:

EAL: using IOMMU type 1 (Type 1)

This gets printed in the DPDK code in function vfio_set_iommu_type() (dpdk/lib/librte_eal/linuxapp/eal/eal_vfio.c).  If this gets printed, it means that the VFIO_SET_IOMMU ioctl() succeeded with VFIO_TYPE1 and that an IOMMU is present on this platform.

So it might be worth trying to remove NOIOMMU=y from your kernel config, but it doesn’t look like that is the problem here.  It looks like there really is an IOMMU but for some platform-specific reason, vfio is not able to map memory for DMA to this device.

The lspci –t data was somewhat garbled – specifically all of the indenting got stripped out by markdown.  Could you put markdown tags around it?  Not sure if you can do this by e-mail or it needs to be done directly from the GitHub website.

Trying it here in e-mail (with three tildes):

~~~
test
                test indent one tab
                                test indent two tabs
~~~

**Comment by ksingh2017** [comment]
![lspci_t](https://user-images.githubusercontent.com/35547979/35162026-dc03b48c-fd39-11e7-86c1-71bc39658800.png)

Jim,
I took a screenshot of the lspci -t output and attached it here.  I hope, the attachment works.

Ziye,
I'll try binding to uio and see if that works.

Thanks,
Kamaljit

**Comment by ksingh2017** [comment]
I forced binding with uio by modifying spdk/scripts/setup.sh.  nvmf_tgt seems to have started without errors but neither discovery nor connect are working.  Looks like vfio support is still being initialized.  Is there a way to disable that?

# nvmf_tgt startup
Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid11999 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
app.c: 369:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 668:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: *NOTICE*: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: *NOTICE*: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 409:spdk_nvmf_construct_subsystem: *NOTICE*: Attaching block device Nvme0n1 to subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: *NOTICE*: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: *NOTICE*: Acceptor running

# Discovery
[qelr_poll_cq_req:1668]Error: POLL CQ with RDMA_CQE_REQ_STS_REMOTE_INVALID_REQUEST_ERR. QP icid=0x1
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876662640 (9): remote invalid request error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672896 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672992 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673088 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673184 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673280 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673376 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673472 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673568 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673664 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673760 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673856 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673952 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674048 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674144 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674240 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674336 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674432 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674528 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674624 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674720 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674816 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674912 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876675008 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876675104 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876675200 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672224 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672320 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672416 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672512 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672608 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672704 (5): Work Request Flushed Error
[qelr_poll_cq_req:1620]Error: POLL CQ with ROCE_CQE_REQ_STS_WORK_REQUEST_FLUSHED_ERR. QP icid=0x1
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672800 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876662640 (5): Work Request Flushed Error

# Connect
[qelr_poll_cq_req:1668]Error: POLL CQ with RDMA_CQE_REQ_STS_REMOTE_INVALID_REQUEST_ERR. QP icid=0x1
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876662480 (9): remote invalid request error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672736 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672832 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672928 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673024 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673120 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673216 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673312 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673408 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673504 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673600 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673696 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673792 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673888 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876673984 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674080 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674176 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674272 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674368 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674464 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674560 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674656 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674752 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674848 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876674944 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876675040 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672064 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672160 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672256 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672352 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672448 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672544 (5): Work Request Flushed Error
[qelr_poll_cq_req:1620]Error: POLL CQ with ROCE_CQE_REQ_STS_WORK_REQUEST_FLUSHED_ERR. QP icid=0x1
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876672640 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x343da1f0, Request 0x876662480 (5): Work Request Flushed Error

**Comment by jimharris** [comment]
Usually it involves disabling the IOMMU in BIOS.

**Comment by ksingh2017** [comment]
There was no IOMMU option in the BIOS.  I turned it off in the kernel config and rebuilt it.  There is no VFIO/IOMMU in the picture now.  nvmf_tgt starts normally but sending a discover command shows these errors on the target.

Interestingly, I saw the same issue when I bound the SSD to uio_pci_generic instead of vfio_pci.

# scripts/setup.sh 
ls: cannot access /sys/kernel/iommu_groups: No such file or directory
0006:58:00.0 (8086 0953): nvme -> uio_pci_generic

# cd app/nvmf_tgt/ 
# ./nvmf_tgt -c nv1_416.conf 
Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid23227 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
app.c: 369:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 668:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2014-08.org.nvmexpress.discovery
nvmf_tgt.c: 108:nvmf_tgt_create_subsystem: *NOTICE*: allocated subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1136:spdk_nvmf_rdma_create: *NOTICE*: *** RDMA Transport Init ***
rdma.c:1349:spdk_nvmf_rdma_listen: *NOTICE*: *** NVMf Target Listening on 192.168.2.150 port 4420 ***
conf.c: 409:spdk_nvmf_construct_subsystem: *NOTICE*: Attaching block device Nvme0n1 to subsystem nqn.2016-06.io.spdk:cnode1
rdma.c:1490:spdk_nvmf_rdma_poll_group_create: *NOTICE*: Skipping unused RDMA device when creating poll group.
nvmf_tgt.c: 302:nvmf_tgt_advance_state: *NOTICE*: Acceptor running



[qelr_poll_cq_req:1668]Error: POLL CQ with RDMA_CQE_REQ_STS_REMOTE_INVALID_REQUEST_ERR. QP icid=0x1
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064029440 (9): remote invalid request error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064039696 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064039792 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064039888 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064039984 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064040080 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064040176 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064039504 (5): Work Request Flushed Error
[qelr_poll_cq_req:1620]Error: POLL CQ with ROCE_CQE_REQ_STS_WORK_REQUEST_FLUSHED_ERR. QP icid=0x1
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064039600 (5): Work Request Flushed Error
rdma.c:1766:spdk_nvmf_rdma_poller_poll: *ERROR*: CQ error on CQ 0x3f68a210, Request 0x1064029440 (5): Work Request Flushed Error

**Comment by ksingh2017** [comment]
Previous message was based on this commit.
commit c3b313811761956dde1d5a2df50b35eec178955d
Author: Xiaodong Liu <xiaodong.liu@intel.com>
Date:   Tue Dec 12 04:53:18 2017 -0500


Since then I've resync'd from git but I still see the same errors.  I'm now using this commit.
commit 8f9028deb3513e2841c8f7b4fedeb2a36d63fd6c
Author: Shuhei Matsumoto <shuhei.matsumoto.xt@hitachi.com>
Date:   Wed Jan 24 15:57:10 2018 +0900

**Comment by ksingh2017** [comment]
Jim,

Can you please look into changing that spdk/lib/env_dpdk/env_internal.h macro  SHIFT_128TB from 47 to 48 like you suggested?  Without that I'm still seeing the following errors even with commit 8f9028d from today Jan 24, 2018.
```
Starting DPDK 17.11.0 initialization...
[ DPDK EAL parameters: nvmf -c 0x1 --file-prefix=spdk_pid42177 ]
EAL: Detected 48 lcore(s)
EAL: No free hugepages reported in hugepages-2048kB
EAL: Probing VFIO support...
**invalid usermode virtual address 0xffff40000000
invalid usermode virtual address 0xfffee0000000**
app.c: 373:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 668:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
reactor.c: 424:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
copy_engine_ioat.c: 306:copy_engine_ioat_init: *NOTICE*: Ioat Copy Engine Offload Enabled
EAL: PCI device 0006:58:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
**invalid usermode virtual address 0xffff0e923000**
**nvme_pcie.c:1642:nvme_pcie_prp_list_append: *ERROR*: vtophys(0xffff0e923000) failed**
nvme_qpair.c: 112:nvme_admin_qpair_print_command: *NOTICE*: IDENTIFY (06) sqid:0 cid:95 nsid:0 cdw10:00000001 cdw11:00000000
nvme_qpair.c: 283:nvme_qpair_print_completion: *NOTICE*: INVALID FIELD (00/02) sqid:0 cid:95 cdw0:0 sqhd:0000 p:0 m:0 dnr:1
bdev.c: 593:spdk_bdev_initialize: *ERROR*: bdev modules init failed
subsystem.c: 122:spdk_subsystem_init_next: *ERROR*: Init subsystem bdev failed
```

**Comment by ksingh2017** [comment]
BTW, during spdk compilation, I'm seeing these 3 errors.  nvmf_tgt is still building though.  These errors seem to be for some test programs. Could they cause any issues for nvmf_tgt?
```
  LINK test/unit/include/spdk/histogram_data.h/histogram_ut
  LINK test/lib/bdev/bdevio/bdevio
  CC app/nvmf_tgt/nvmf_tgt.o
  CC test/unit/lib/bdev/bdev.c/bdev_ut.o
  CC examples/ioat/verify/verify.o
iscsi_top.cpp:36:10: fatal error: algorithm: No such file or directory
 #include <algorithm>
          ^~~~~~~~~~~
trace.cpp:36:10: fatal error: map: No such file or directory
 #include <map>
          ^~~~~
compilation terminated.
compilation terminated.
make[2]: make[2]: *** [trace.o] Error 1*** [iscsi_top.o] Error 1

make[1]: *** [trace] Error 2
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [iscsi_top] Error 2
  CC test/unit/lib/bdev/scsi_nvme.c/scsi_nvme_ut.o
  CC examples/blob/cli/blobcli.o
  CC examples/nvme/identify/identify.o
  CC test/app/bdev_svc/bdev_svc.o
  CXX test/cpp_headers/net.o
  CC examples/nvme/perf/perf.o
  CXX test/cpp_headers/copy_engine.o
  CC app/nvmf_tgt/nvmf_rpc.o
  CXX test/cpp_headers/iscsi_spec.o
  LINK test/app/bdev_svc/bdev_svc
  CC test/lib/bdev/bdevperf/bdevperf.o
  LINK examples/ioat/verify/verify
  LINK test/unit/lib/bdev/scsi_nvme.c/scsi_nvme_ut
  CXX test/cpp_headers/io_channel.o
  CC examples/ioat/kperf/ioat_kperf.o
  LINK app/nvmf_tgt/nvmf_tgt
  CC test/unit/lib/bdev/vbdev_lvol.c/vbdev_lvol_ut.o
  CC test/unit/lib/bdev/gpt/gpt.c/gpt_ut.o
  LINK examples/blob/cli/blobcli
  CXX test/cpp_headers/lvol.o
  CC test/unit/lib/bdev/mt/bdev.c/bdev_ut.o
  LINK test/lib/bdev/bdevperf/bdevperf
  LINK examples/ioat/kperf/ioat_kperf
  LINK examples/nvme/identify/identify
  CC test/unit/lib/blob/blob.c/blob_ut.o
  LINK examples/nvme/perf/perf
  CC test/unit/lib/blobfs/tree.c/tree_ut.o
make: *** [app] Error 2
make: *** Waiting for unfinished jobs....
  CC test/unit/lib/event/subsystem.c/subsystem_ut.o
  CXX test/cpp_headers/queue.o
  CC test/unit/lib/ioat/ioat.c/ioat_ut.o
  CXX test/cpp_headers/trace.o
  LINK test/unit/lib/bdev/gpt/gpt.c/gpt_ut
  CXX test/cpp_headers/ioat_spec.o
  CC examples/nvme/reserve/reserve.o
  CC test/lib/blobfs/blobfs_async_ut/blobfs_async_ut.o
  CC test/lib/blobfs/blobfs_sync_ut/blobfs_sync_ut.o
  LINK test/unit/lib/bdev/bdev.c/bdev_ut
blob_ut.c: In function ?bs_version?:
blob_ut.c:2423:55: error: ?PAGE_SIZE? undeclared (first use in this function)
  memset(&g_dev_buffer[super->used_blobid_mask_start * PAGE_SIZE], 0,
                                                       ^
blob_ut.c:2423:55: note: each undeclared identifier is reported only once for each function it appears in
  LINK test/unit/lib/blobfs/tree.c/tree_ut
  CC test/lib/blobfs/mkfs/mkfs.o
make[5]: *** [blob_ut.o] Error 1
make[4]: *** [blob.c] Error 2
make[3]: *** [blob] Error 2
make[3]: *** Waiting for unfinished jobs....
  CC test/lib/env/memory/memory_ut.o
  CC examples/nvme/nvme_manage/nvme_manage.o
  LINK test/unit/lib/event/subsystem.c/subsystem_ut
  CC test/lib/env/pci/pci_ut.o
  LINK examples/nvme/reserve/reserve
  CXX test/cpp_headers/conf.o
  CXX test/cpp_headers/bit_array.o
  LINK test/unit/lib/ioat/ioat.c/ioat_ut
  LINK test/lib/blobfs/mkfs/mkfs
  LINK test/unit/lib/bdev/vbdev_lvol.c/vbdev_lvol_ut
  CXX test/cpp_headers/string.o
  CXX test/cpp_headers/gpt_spec.o
  CXX test/cpp_headers/queue_extras.o
  LINK test/lib/env/pci/pci_ut
  CC examples/nvme/arbitration/arbitration.o
  CC examples/nvme/hotplug/hotplug.o
  CXX test/cpp_headers/rpc.o
  CC test/lib/env/vtophys/vtophys.o
  LINK examples/nvme/nvme_manage/nvme_manage
  CXX test/cpp_headers/ioat.o
  LINK test/lib/env/memory/memory_ut
  CC test/lib/event/event_perf/event_perf.o
  CC test/lib/event/reactor/reactor.o
  CC test/lib/event/reactor_perf/reactor_perf.o
  LINK test/unit/lib/bdev/mt/bdev.c/bdev_ut
  LINK test/lib/env/vtophys/vtophys
make[2]: *** [lib] Error 2
make[1]: *** [unit] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by jimharris** [comment]
For <map> and <algorithm> - can you confirm you’ve either run the pkgdep.sh script or installed all of the packages listed there?  If so – then there must be another package on your system that provides C++ STL headers that needs to be added to pkgdep.sh.

I’ll add a patch to fix the PAGE_SIZE error.

These should cause no problems for nvmf_tgt though.

-Jim

**Comment by jimharris** [comment]
https://review.gerrithub.io/#/c/396320/ should fix the PAGE_SIZE error on your system.

**Comment by ksingh2017** [comment]
>https://review.gerrithub.io/#/c/396320/ should fix the PAGE_SIZE error on your system.
Yes, that's fixed now.  Thanks!

-Kamaljit

**Comment by jimharris** [comment]
Thanks for the confirmation.  These patches have been pushed to master.  Let us know what you find on the C++ STL include errors.

-Jim

**Comment by ksingh2017** [comment]
Running pkgdep.sh only installed sg3_utils.  Looks like I already had most of the other dependencies installed. It didn't find the following packages but these don't look relevant for this issue.
No package libpmemblk-devel available.
No package mscgen available.
No package astyle-devel available.

So, those 2 errors still exist but it looks like an env issue.  You can ignore it for now.

Thanks,
-Kamaljit

**Comment by ksingh2017** [comment]
Jim, Ben and Ziye,
I'm opening another issue to track the discover/connect issue on ARM64.  Ignoring iommu/vfio for now.

Thanks for your help so far.

-Kamaljit

**Comment by liang2yan** [comment]
@ksingh2017 , can we close this issue?

**Comment by jimharris** [comment]
Closing this issue.  @ksingh2017, if you feel there is still something to look at here, please reopen it.

Thanks,

-Jim
